Start adding GRW RV Op and distribution

canyon289 opened the pull request:
Starting to add time series support to v4 on step at a time (See what I did there?)

Scope of this PR is just to add the random variable, not the distribution as this might take me a while and I'll likely have a lot of basic questions

https://github.com/pymc-devs/pymc/issues/5290

ricardoV94 commented:
Code inside the `rng_fn` should be pure non-symbolic numpy/python.

Anyway the idea with aeppl is that we don't need to build a new RandomVariable for every possible distribution, but use Aesara directly (as you are doing with that scan) to generate random draws. 

Similarly, instead of needing a custom logp method for every distribution we use aeppl to generate the logp graph, given the corresponding Aesara random generating graph.

The thing that is needed is to make `pm.GaussianRandomWalk` return the right Aesara random generating graph. We still have to figure out the right API for this type of "distributions'. I have a WIP proposal in #5169 under the name of `DerivedDistribution` (feedback and better name appreciated)

ricardoV94 commented:
Also to illustrate the point of this approach, the only thing that is needed to create any other random walk is to change the distribution used in that scan graph

```python
rv, updates = aesara.scan(
  fn=lambda prev_value: pm.StudentT.dist(prev_value + mu, sigma, nu),
  outputs_info=np.array(0.0),
  n_steps=size,
)
# ignoring: updates have to be managed differently when using RandomVariables via `.dist`
```

Instead of having to create them one by one we can offer all at once in a `pm.RandomWalk` by letting users specify which base distribution to use as one of the arguments.

```python
my_srw = pm.RandomWalk("my_srw", dist=pm.StudentT.dist(mu, sigma, nu), size=10)
```

 `pm.GaussianRandomWalk` can then be a short convenience function that calls `pm.RandomWalk(name, dist=pm.Normal.dist(mu, sigma), size=size)` under the hood.

canyon289 commented:
These are great pointers thank you!

canyon289 commented:
@ricardoV94, reviewed the other prs and thought through this some. Do you mind if I learn how to implement a "fixed" GRW RV first in this PR, get that working, then follow up with a RW RV generalization? Thatll help me break up unknown bits that I need to learn into smaller unknown bits for me

ricardoV94 commented:
> @ricardoV94, reviewed the other prs and thought through this some. Do you mind if I learn how to implement a "fixed" GRW RV first in this PR, get that working, then follow up with a RW RV generalization? Thatll help me break up unknown bits that I need to learn into smaller unknown bits for me

If you want to take it in small steps I think limiting yourself to the case of Gaussian innovations might help. I would still suggest you sidestep writing the RandomVariable altogether and try to use the WIP `DerivedDistribution` from my branch.

Otherwise you'll just be practicing writing traditional PyMC distributions. That can be useful for temporary backwards compatibility of V4, but it will have a smaller impact than understanding how to make use of aeppl's flexibility.

canyon289 commented:
Great thanks for this advice, I'll start working off of that branch

canyon289 commented:
Why is precommit failing here? Same commit, different results locally than in Ci?

![image](https://user-images.githubusercontent.com/7213793/147882127-542403a7-b479-4610-bc52-ef9a4e087b3b.png)

canyon289 commented:
@ricardoV94 Let me know if you like this GRW RV implementation.

Per our conversation for now we'll
1. Skip init distribution
2. Restrict innovations just to Gaussians to get something working
3. Implement an old style distribution


Next steps for me to to create the GRW Distribution in old style

After thats working I'll work on
1. Switching to Derived dist with more aeppl and aesara integration
2. Generalize distributions
3. Figure out init distribution (if we want it)

ricardoV94 commented:
Are you planning to do this all in one PR or separate ones?

canyon289 commented:
> Are you planning to do this all in one PR or separate ones?

Separate ones ideally

i updated the comment to make it more clear what my plan is, let me know if you think its a reasonable course of action

ricardoV94 commented:
Are you planning to refactor the PyMC distribution to work with the explicit RV from this PR?

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +36,55 @@
+        mu: Union[np.ndarray, float],
+        sigma: Union[np.ndarray, float],
+        init: float,
+        size: int,
```
Currently the tests are failing because the size argument provided is not recognized in the perform op. Another size is created somewhere else it seems, leading to the following error

> E       TypeError: rng_fn() takes 6 positional arguments but 7 were given

Here's the closet line where I notice things going wrong.
https://github.com/aesara-devs/aesara/blob/main/aesara/tensor/random/op.py#L363

Any idea of where I'm going wrong? I'm following the guide below as well as the `TruncatedNormal` implementation

https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html
https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L601
![image](https://user-images.githubusercontent.com/7213793/148153891-958e5cbb-a7b6-4a20-bc84-37e53bc80a66.png)

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -17,15 +17,33 @@
+    sd = 0.0000001
+    size = 4
+
+    grw = gaussianrandomwalk(mu, sd, init, size).eval()
```
Here is direct link to testing exception where the size argument ends up in the wrong place. Any advice would be great!

https://github.com/pymc-devs/pymc/runs/4710136699?check_suite_focus=true#step:7:2295

ricardoV94 replied:
I think size is a keyword only argument in RandomVariables

canyon289 replied:
That was it thank you

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -201,7 +253,25 @@ class GaussianRandomWalk(distribution.Continuous):
+    @classmethod
+    def dist(
+        cls,
+        mu: Optional[Union[np.ndarray, float]] = None,
```
These type hints don't consider the case when inputs are other Aesara variables

canyon289 replied:
Got it, I copied them from TruncatedNormal. Are these ones wrong too? If so I can update them too

https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L601

ricardoV94 replied:
`int` are also valid. Because there's so many we should probably save the union of these type hints in a variable that we can reuse a across distributions

ricardoV94 replied:
> Got it, I copied them from TruncatedNormal. Are these ones wrong too? If so I can update them too
> 
> https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L601

Yes they are also incomplete

canyon289 replied:
Ok ill open up another PR just for this to catch up the distributions on main, and then implement it here too of course

canyon289 replied:
Started on this PR https://github.com/pymc-devs/pymc/pull/5315

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -237,64 +298,13 @@ def logp(self, x):
-        else:
-            data = data - data[0]
-        return data
+        rv, updates = aesara.scan(
```
Alright heres my question, how do I properly implement a the logp method? Here's why I'm confused on this question

1. The [distribution docs](https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html#inheriting-from-a-pymc-base-distribution-class) and other distributions such as [Kumarswamy](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1359) show calculating the logp totally "manually"
2. Some distributions like [Exponential](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1420) don't have a logp at all, my guess is because the rv_op is defined in Aesara?
3. The pretty cool colab that Ricardo put together [uses scan](https://gist.github.com/ricardoV94/a4da531a60a0f16facf8dbb5416a2cd4#file-aeppl_scan-py-L22)

Out of the three approaches I'm not entirely sure which one should be used here, as well as some general confusion about the logp"less" distributions

ricardoV94 replied:
> Alright heres my question, how do I properly implement a the logp method? Here's why I'm confused on this question
> 
>     1. The [distribution docs](https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html#inheriting-from-a-pymc-base-distribution-class) and other distributions such as [Kumarswamy](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1359) show calculating the logp totally "manually"
> 
>     2. Some distributions like [Exponential](https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/continuous.py#L1420) don't have a logp at all, my guess is because the rv_op is defined in Aesara?
> 
>     3. The pretty cool colab that Ricardo put together [uses scan](https://gist.github.com/ricardoV94/a4da531a60a0f16facf8dbb5416a2cd4#file-aeppl_scan-py-L22)
> 
> 
> Out of the three approaches I'm not entirely sure which one should be used here, as well as some general confusion about the logp"less" distributions

1. and 3. The manual implementation is used for "standard" distributions (which have a specialized RandomVariable, like the one you implemented here), which you wanted to try as the first step. The point of later switching to Aeppl is that it will compute it for you, so there is no need to write it by hand (neither the RandomVariable nor the logp expression).
2.  Those are implemented in Aeppl here: https://github.com/aesara-devs/aeppl/blob/333108143a1f4b63d9fcd9842dbe35457025c180/aeppl/logprob.py#L191-L198 (we should probably leave a comment in our codebase so that devs/users can find it more easily)

ricardoV94 replied:
Here something like the following should be enough (haven't tested it!):

```python
innit_logp = pm.logp(pm.Normal.dist(mu, sigma), value[:1] - init)
innov_logp = pm.logp(pm.Normal.dist(mu, sigma), at.diff(value))
total_logp = at.concatenate([innit_logp, innov_logp])
return total_logp
```

No need for a scan for the case of GaussianRandomWalk logp

ricardoV94 replied:
You shouldn't call aeppl inside a logp expression, because aeppl actually needs the pairs `{random_variable: value_variable}` for all the variables involved in the logp graph (`value`, `mu`, `sigma`, `init`), but when `GaussianRandomWalk.logp` is called it's only given the `random_variables` for those.

canyon289 replied:
Thanks for all the advice, I'm going give it a shot today.

> but when GaussianRandomWalk.logp is called it's only given the random_variables for this

Im confused on point, the docs show `logp` taking `value` and `param1`?
https://docs.pymc.io/en/latest/contributing/developer_guide_implementing_distribution.html
<img width="664" alt="image" src="https://user-images.githubusercontent.com/7213793/148655150-68fdd5af-7d90-4ec5-ae67-877bb4c9f298.png">

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -201,30 +252,40 @@ class GaussianRandomWalk(distribution.Continuous):
+    # TODO: Add typehints
+    def get_moment(
+        self,
+        size: Optional[Union[np.ndarray, float]],
```
These can (and are almost always) symbolic inputs

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -201,30 +252,40 @@ class GaussianRandomWalk(distribution.Continuous):
+        sigma: Optional[Union[np.ndarray, float]],
+        init: Optional[Union[np.ndarray, float]],
+    ):
+        moment = mu * size + init
```
when size is none this would fail

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -201,30 +252,40 @@ class GaussianRandomWalk(distribution.Continuous):
+        return moment
+
+    def logp(
+        value,
```
Also almost always symbolic inputs

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -237,64 +297,39 @@ def logp(self, x):
+        innit_logp = normal_logp(value[0] - init, mu, sigma)
+
+        # Create logp calculation graph for innovations
+        stationary_vals = at.diff(value[1:]) - init
```
I don't think `init` matters anymore here. Also I think the `diff` should include the whole array. I could be mistaken

canyon289 replied:
youre right, init is not needed. 

Diff returns an array thats len(n) -1, fixed the `[1:]` indexing as such

![image](https://user-images.githubusercontent.com/7213793/149548784-9ea32bed-a964-4f84-a801-9fc9ef34343f.png)

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -237,64 +297,39 @@ def logp(self, x):
-        return ["mu", "sigma"]
+        # Return both calculation logps in a vector. This is fine because somewhere
+        # down the line these will be summed together
+        total_logp = at.concatenate([innit_logp, innov_logp])
```
This is not only fine but needed, for example for model comparison where you need the elemwise likelihood of the observations. Thats why our logp methods dont reduce the result to scalars by default

canyon289 replied:
Got it thanks

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -237,64 +297,39 @@ def logp(self, x):
+        # down the line these will be summed together
+        total_logp = at.concatenate([innit_logp, innov_logp])
+
+        return total_logp
```
should wrap this in a `CheckParameterValues` that checks for positive sigma

canyon289 replied:
Added, Thanks

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -201,30 +253,39 @@ class GaussianRandomWalk(distribution.Continuous):
+    #     return moment
+
+    def logp(
+        value: Union[at.TensorVariable, at.TensorConstant],
```
Did I get this type hint correct now? It was noted that it "almost" always is a symbolic input. Could it be something else?

ricardoV94 replied:
I think you can use the more generic `Variable` which covers both types.

ricardoV94 replied:
> Could it be something else?

It should not

canyon289 replied:
To be really precise here, should it not, or is it not? I ask to ensure the type hint is correct. Updated to at.TensorVariable in meantime

ricardoV94 replied:
> To be really precise here, should it not, or is it not? I ask to ensure the type hint is correct. Updated to at.TensorVariable in meantime

I mean `aesara.graph.basic.Variable` or something like that which works for both `TensorVariables` and `TensorConstants` I think. But maybe TensorConstants are now a subclass of TensorVariables? I'll have to check... Actually running mypy would help a lot with all this guess and recalling work xD

The "should not" question, it is because if users call it directly with numpy arrays it will more often than not work perfectly fine. But internally we always pass symbolic variables and that's what we test in our suite.

canyon289 replied:
hm 🤔 
Here's my proposal then. Lets keep it `Union[at.TensorVariable, at.TensorConstant]` Here's my thinking

The things users will see they inspect types `at.Tensorvariable` or `at.TensorConstant`. Even though it would be terser to use the common parent class it would require impicit knowledge of all subclasses that come from that which im not sure that people have

+1 to mypy checks as well. After I get this PR merged I can work on pushing that along.

ricardoV94 replied:
Yeah I agree with you. Seems more readable

canyon289 replied:
:tada:

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -17,15 +17,51 @@
+    with pm.Model():
+        grw = GaussianRandomWalk("grw", mu, sd, init, size=2)
+
+    logp = pm.logp(grw, vals)
```
Alright @ricardoV94 Im at the next stage of development challenge which is shape issue.

I get what the error is saying but I'm not sure exactly what to do about it. My questions in order is
1. is the way I'm writing the test correct e.g. Should I be testing through the pm.logp api?
2. What is the fix to the shape issue do you think?

```
 self = <aesara.tensor.basic.Join object at 0x7ff3b387ca90>, axis = 0
tens = (Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0)
as_tensor_variable_args = [Elemwise{sub,no_inplace}.0, Elemwise{sub,no_inplace}.0]
output_maker = <function Join.make_node.<locals>.output_maker at 0x7ff39e1d04d0>

    def _make_node_internal(self, axis, tens, as_tensor_variable_args, output_maker):
        if not builtins.all(targs.type.ndim for targs in as_tensor_variable_args):
            raise TypeError(
>               "Join cannot handle arguments of dimension 0."
                " For joining scalar values, see @stack"
            )
E           TypeError: Join cannot handle arguments of dimension 0. For joining scalar values, see @stack
```

https://github.com/pymc-devs/pymc/runs/4825257275?check_suite_focus=true#step:7:2040

ricardoV94 replied:
I replied below with a suggestion for the shape issue. The logp you should test like we do for normal distributions, creating the distribution in a model context and evaluating the model logp for a couple of points. 

I would take a look at the Multinomial tests we have in `test_distributions.py` to have an idea of the range of logp tests we have for multivariate distributions.

Also, didn't we have any previous tests for the GRW? That's surprising

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -237,64 +298,27 @@ def logp(self, x):
 
-    def random(self, point=None, size=None):
-        """Draw random values from GaussianRandomWalk.
+        def normal_logp(value, mu, sigma):
```
Implementing in this janky way for now just to learn "from scratch"

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -237,64 +298,27 @@ def logp(self, x):
-            Desired size of random sample (returns one sample if not
-            specified).
+        # Create logp calculation graph the inital time point
+        init_logp = normal_logp(value[0] - init, mu, sigma)
```
Your shape issues probably come from here. You will end up trying to concatenate the scalar init logp with the vector innov logp. If you instead select the initial value like this, it should work:

```suggestion
        init_logp = normal_logp(value[:1] - init, mu, sigma)
```

Alternatively you can use the strategy we use in aeppl for cumsum "SymbolicDists", where we first concatenate the initial value with the vector of differences and only then call the logp: https://github.com/aesara-devs/aeppl/blob/d09a01d8f165c709be0b73b51d5a8725d91d7d94/aeppl/cumsum.py#L20

(That code is a bit more complicated because we handle cumsum across any axis, you only need to worry about `axis=-1` since thats what the GRW Op uses).

canyon289 replied:
Thanks!

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -179,122 +237,68 @@ def logp(self, value):
-        else:
-            data = data - data[0]
-        return data
+        total_logp = at.concatenate([at.expand_dims(init_logp, 0), series_logp])
```
Used `expand_dims` rather than `[:1]` to make the dimension expansion explicit rather than implicit

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -179,122 +237,68 @@ def logp(self, value):
-            data = data - data[0]
-        return data
+        total_logp = at.concatenate([at.expand_dims(init_logp, 0), series_logp])
+        total_logp = check_parameters(total_logp, sigma > 0, msg="sigma > 0")
```
In this case you don't need the `check_parameters`, since it is added by the `pm.logp` call above already

ricardoV94 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -15,17 +15,64 @@
-# pytestmark = pytest.mark.usefixtures("seeded_test")
-pytestmark = pytest.mark.xfail(reason="Timeseries not refactored")
+
+def test_grw_rv_op():
```
Looks good, we should also test 2d sizes and values to make sure both the rng and logp are working with the correct axis

ricardoV94 replied:
There is also an old test here that's should be removed: https://github.com/pymc-devs/pymc/blob/d52655deb97303b68b8d2e7b2fdb2b9229ee67af/pymc/tests/test_distributions_random.py#L252

ricardoV94 replied:
Alternatively, you may also want to borrow the standardized RV op tests in `test_distributions_random.py` (Or perhaps more logically put the RV test there)

canyon289 replied:
To understand this more precisely should I
1. Reenable this current test get all tests to pass?
2. Remove this one entirely, but construct new ones to replace it?

Asking as without a clear understanding of correct implementation shape, steps/length, size not sure what the best route is

Here's all the tests that are failing
<img width="428" alt="image" src="https://user-images.githubusercontent.com/7213793/150703963-8736e8fd-e282-4527-9eaa-12f3bbf0ad5b.png">

canyon289 replied:
on further inspection no other test class seems to inherit from `BaseTestCases.BaseTestCase`, so I'm proposing we remove it. That being said I wouldnt know what replace it with for GRW case at the moment


https://github.com/pymc-devs/pymc/blob/d52655deb97303b68b8d2e7b2fdb2b9229ee67af/pymc/tests/test_distributions_random.py#L156

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +37,64 @@
+        mu: Union[np.ndarray, float],
+        sigma: Union[np.ndarray, float],
+        init: float,
+        length: int,
```
Maybe call this steps? Also you will have to implement a custom `_shape_from_params` so that Aesara can know the expected shape of the RV

canyon289 commented:
Thanks @ricardoV94 for the feedback

fonnesbeck commented on pymc/distributions/timeseries.py:
```diff
@@ -12,11 +12,15 @@
-from scipy import stats
+from aesara.tensor.random.basic import RandomVariable
+
+import pymc as pm
```
Replace this with specific imports that you need.

canyon289 replied:
Will do. For reference I have this same confusion for the entire library. We use `import pymc as pm` in many other submodules

https://github.com/pymc-devs/pymc/blob/main/pymc/distributions/multivariate.py#L42-L53
https://smartinference.slack.com/archives/C6AKTNSBS/p1642434564015100

canyon289 replied:
done

fonnesbeck commented on pymc/distributions/timeseries.py:
```diff
@@ -179,122 +241,66 @@ def logp(self, value):
+        return super().dist([mu, sigma, init], **kwargs)
+
+    def logp(
+        value: at.Variable,
```
Shouldn't the types be ndarray or TensorVariable?

canyon289 replied:
Per discussion even though it could it sounds like its not something that is advisable https://github.com/pymc-devs/pymc/pull/5298#discussion_r785276361

fonnesbeck replied:
But a numpy ndarray will get used quite a lot here, no? Seems like a potential point of confusion for end users.

ricardoV94 replied:
logps are called with the inputs of the respective RandomVariable so they should always be symbolic, even if the user created the distribution with numpy arrays (in make_node they are converted to tensor_variables automatically)

canyon289 replied:
Thanks @ricardoV94 :). This is part of the reason I really like that adding type hints, as it helps clear up my confusion as well!

mjhajharia commented:
thanks for this @canyon289 !! minor nitpick, Numpy uses default_rng() now for random generator I think instead of random state, its just a one line replacement i think https://numpy.org/doc/stable/reference/random/generator.html

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +37,64 @@
+    """
+
+    name = "GaussianRandomWalk"
+    ndim_supp = 0
```
```suggestion
    ndim_supp = 1
```

canyon289 commented:
> 

> thanks for this @canyon289 !! minor nitpick, Numpy uses default_rng() now for random generator I think instead of random state, its just a one line replacement i think https://numpy.org/doc/stable/reference/random/generator.html

Thanks for review. Happy to make the change, where should I do it?

canyon289 commented:
More questions if folks don't mind

Here's what the input and outputs for `size`, `steps`, and the returned shape for at last the `rng_fn`
1. First and most broadly, does this make sense design wise? Let's assume 1D GRW and only scalar inputs allowed right now

| size arg | steps arg | output shape |
|----------|-----------|--------------|
| None     | None      | (1,)         |
| 1        | None      | (1,)         |
| None     | 1         | (2,)         |
| 1        | 10        | (11,)        |
| 10       | 1         | (10,2)       |
| 10       | 10        | (10,11)      |


2. Am I on the right track with this `_shape_from_params`?
   https://github.com/pymc-devs/pymc/pull/5298/commits/ee8e349c267066642fcc2109742b8018610cfebb#diff-5fece42aa9647732b360b6c0e13de089a8408c68ae38b650fba38a1c40322229R54
3. The current shape tests are failing, but unsure if that's because the tests are outdated or my code is wrong
https://github.com/pymc-devs/pymc/blob/main/pymc/tests/test_distributions_random.py#L207

ricardoV94 commented:
I think it should be

| size arg | steps arg | output shape |
|----------|-----------|--------------|
| None     | None      | (,)         |
| 1        | None      | (1,)         |
| None     | 1         | (1,)         |
| 1    | 1         | (1,1)         |
| 1        | 10        | (1,10)        |
| 10       | 1         | (10,1)       |
| 10       | 10        | (10,10)      |

And I would maybe error out on steps = None/0, getting rid of the first 2 lines?

canyon289 commented:
> I think it should be
> size arg 	steps arg 	output shape
> None 	None 	(,)
> 1 	None 	(1,)
> None 	1 	(1,)
> 1 	1 	(1,1)
> 1 	10 	(1,10)
> 10 	1 	(10,1)
> 10 	10 	(10,10)
> 
> And I would maybe error out on steps = None/0, getting rid of the first 2 lines?

Thank you Ricardo! I'm trying this out. Agree on on `steps is None` and `steps == 0` as those are confusing to handle.

![image](https://user-images.githubusercontent.com/7213793/151493822-cdebd00d-425a-4570-b346-0e74a690092b.png)

canyon289 commented:
Here's what I landed on. Like in regression where we have slope and intercept, i ended up figuring we'd want to have an estimatable initialization value, and an estimated  grw walk with n_steps from there.

Assuming there is no issue with the `rng_fn` shape all I need to do now is complete the `_shape_from_params`method right? Or is there something more?

|                  Code                  |  Output |
|:--------------------------------------:|:-------:|
| print(rng_fn(size=None,steps=1).shape) |   (2,)  |
|   print(rng_fn(size=1,steps=1).shape)  |   (2,)  |
|  print(rng_fn(size=1,steps=10).shape)  |  (11,)  |
|   print(rng_fn(size=3,steps=1).shape)  |  (3, 2) |
|  print(rng_fn(size=3,steps=10).shape)  | (3, 11) |


Reference code in case it helps anyone
```
import numpy as np
rng = np.random.default_rng()


def rng_fn(size, steps, rng=rng, init=0, mu=1, sigma=1):
    if steps is None or steps == 0:
        raise ValueError("Steps must be greater than 0 or not None")
    if size is None:
        size = 1

    init_val = rng.normal(init, sigma, size=(size,1))
    steps = rng.normal(loc=mu, scale=sigma, size=(size, steps))

    grw = np.concatenate([init_val, steps], axis=-1)
    return np.cumsum(grw,axis=-1).squeeze()

# Let's ignore none case
print(rng_fn(size=None,steps=1).shape)
print(rng_fn(size=1,steps=1).shape)
print(rng_fn(size=1,steps=10).shape)
print(rng_fn(size=3,steps=1).shape)
print(rng_fn(size=3,steps=10).shape)
```

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +37,93 @@
+            size = 1
+
+        init_val = rng.normal(init, sigma, size=(size, 1))
+        steps = rng.normal(loc=mu, scale=sigma, size=(size, steps))
```
This is all a bit confusing to me, but would this make more sense?
```suggestion
        steps = rng.normal(loc=mu, scale=sigma, size=(size, steps-1))
```

canyon289 replied:
I see you resolved this.  Does this makes sense now?

ricardoV94 commented:
Sounds reasonable, but size `1` should be a `(1, steps)` array, whereas size `None` should be a `(steps,)` array. That's a pretty strong convention in other distributions/Numpy

canyon289 commented:
> Sounds reasonable, but size `1` should be a `(1, steps)` array, whereas size `None` should be a `(steps,)` array. That's a pretty strong convention in other distributions/Numpy

Thanks! I didn't know that. I updated the code. Here's the updated output. Let me know if theres any issues

```
print(rng_fn(size=None,steps=1).shape)
print(rng_fn(size=1,steps=1).shape)
print(rng_fn(size=1,steps=10).shape)
print(rng_fn(size=3,steps=1).shape)
print(rng_fn(size=3,steps=10).shape)
```

> (2,)
> (1, 2)
> (1, 11)
> (3, 2)
> (3, 11)

Stepping back from code details for a second and looking over this general PR let me know if this checklist is correct.

- [x] Implement a custom `_shape_from_params` (will be asking for help here)
- [ ] Update tests
  - [x] Random Method shape test
  - [x] LogP correctness test
  - [ ] LogP shape test
  - [ ] Ad hoc inference test
- [ ] Fix up any docstrings

Thanks immensely @ricardoV94

canyon289 commented:
Marking as draft to stop burning CI minutes, though I really do need active help on this one so please keep looking at the code 😅

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -179,122 +279,67 @@ def logp(self, value):
-    def random(self, point=None, size=None):
-        """Draw random values from GaussianRandomWalk.
+        # Calculate initialization logp
+        init_logp = pm.logp(Normal.dist(init, sigma), 0)
```
?

```suggestion
        init_logp = pm.logp(Normal.dist(init, sigma), value[0])
```

canyon289 replied:
Thanks for catching

canyon289 commented:
After much work and help from @ricardoV94 this is ready for review again!

Time series tests are working for me locally, though I'm on an older version of Aesara. Test failures in CI are related to some new bug that was introduced. Nonetheless would like to verify the code here is good, and once we confirm we can figure out what to do with the Aesara blocker.

![image](https://user-images.githubusercontent.com/7213793/154890372-e934121d-b5d8-4212-9ea0-1f333f7b865c.png)

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +35,159 @@
+        **kwargs
+    ) -> RandomVariable:
+
+        params = [at.as_tensor_variable(param) for param in (mu, sigma, init, steps)]
```
We usually wrap parameters in either `floatX` or `intX`. So in this case something like:
```suggestion
        params = [at.as_tensor_variable(floatX(param)) for param in (mu, sigma, init)] + [at.as_tensor_variable(intX(steps))]
```

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +35,159 @@
+            init_size = (*size, 1)
+            steps_size = (*size, steps)
+
+        init_val = rng.normal(init, sigma, size=init_size)
```
I was thinking about this again and I don't think we should sample the `init_val`, we should just use whatever `init` is. The motivation is more clear in the logp... see my comment below

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +35,159 @@
+        sigma: Optional[Union[np.ndarray, float]] = 1.0,
+        init: float = 0.0,
+        steps: int = 0,
+        *args,
```
This may create less confusion, users have to specify steps and init by keyword only. And steps is not optional (since zero is invalid), or we can make the default 1? We can also use `shape[-1]` for steps, if shape is provided (we would do this in `__new__` since `dist` does not accept `shape`, only `size`)...

```suggestion
        mu = 0.0,
        sigma = 1.0,
        *,
        steps: int,
        init = 0.0,
```

The type hints are wrong, because the paramaters can be TensorVariables

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +35,159 @@
+        """
+
+        # Calculate initialization logp
+        init_logp = logprob.logp(Normal.dist(init, sigma), value[0])
```
I don't think we should model the probability of the initial point. Users may want to give this an arbitrary prior, like for mu and sigma. Adding a logp term of this may lead to double counting of the same value and sampling issues. For example when a user does:

```python
with pm.Model() as m:
  init = pm.StudentT("init", 4, 0, 1)
  grw = pm.GaussianRandomWalk("grw", 0, 1, steps=5, init=init)
```

We would be double counting init, once with a StudentT logp and again with the Normal logp.

ricardoV94 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -15,17 +15,103 @@
-# pytestmark = pytest.mark.usefixtures("seeded_test")
-pytestmark = pytest.mark.xfail(reason="Timeseries not refactored")
+
+class TestGaussianRandomWalk:
```
I would suggest reusing the test classes we already have for RandomVariable Ops: https://github.com/pymc-devs/pymc/blob/c892317aba0663df0d250c1813b2ca2fe3460774/pymc/tests/test_distributions_random.py#L263

And for logp: https://github.com/pymc-devs/pymc/blob/c892317aba0663df0d250c1813b2ca2fe3460774/pymc/tests/test_distributions.py#L626

They implement a more extended version of the checks you are doing here (and can be extended). This will make sure we test this distribution at least as thoroughly as we do other distributions, and also that when we  extend those test classes in the future, this benefits automatically as well.

ricardoV94 reviewed (changes requested):
Great progress, I think we are getting to the last touches part of the PR :D

ricardoV94 commented:
> Time series tests are working for me locally, though I'm on an older version of Aesara. Test failures in CI are related to some new bug that was introduced. Nonetheless would like to verify the code here is good, and once we confirm we can figure out what to do with the Aesara blocker.

That bug only applies to cases where the `value` is a constant, and Aesara tries to constant fold the results. I think everything will still work but users will see the "optimization" error whenever they use constant values (as when the distribution is observed). I'll try to patch this some time this week: https://github.com/aesara-devs/aesara/issues/823

In the meantime we can make all tests use symbolic `value`, and add a separate test with an `xfail` mark that uses constant `value`, so that we notice when Aesara is updated and this is no longer a failure.

canyon289 commented:
Thanks for review @ricardoV94. I agree with all the stuff you stated, will get it done

canyon289 commented:
Thanks @ricardoV94 for all the birthday comments, just so you know its still my birthday so youre still socially obligated to leave me more comments on all below

* Init removed
* Added `test_random` method (have some questions here)
* Added `test_distribution` method (definitely have questions here)
- [ ] Remove `test_time_series` test

![image](https://user-images.githubusercontent.com/7213793/155054137-ce563d38-2477-4d5a-bc31-15cbb2c6ca9f.png)
![image](https://user-images.githubusercontent.com/7213793/155054450-f3f75fa0-bae9-4bc6-b843-e192fee2fe46.png)

canyon289 commented on pymc/tests/test_distributions_random.py:
```diff
@@ -417,28 +314,34 @@ def seeded_numpy_distribution_builder(dist_name: str) -> Callable:
+    expected_rv_op_params = {"mu": 1.0, "sigma": 3, "steps": 2}
+    # reference_dist_params = {"b": 1.0, "kappa": 1.0, "mu": 0.0}
+
     checks_to_run = [
```
Should any additional checks be added? You mentioned automatic implied dims. Is this the right place?

ricardoV94 replied:
Same here, I would move this to the test_distributions_timeseries file

canyon289 replied:
Agreed, Ill make an to move this in a future PR just so we can get this merged with minimal headache

ricardoV94 replied:
I don't think it makes sense to add new tests here if we already plan to move them.

For your other question, no need to test implied dims I think.

canyon289 replied:
I dont think we can move these because the test fixture `BaseTestDistributionRandom` is specified in this file, unless we move the whole fixture to conftest.py so its globally available, but I think thats a worse option.

Because its a test class maybe we can move it but even then Id disagree since it adds another layer of indirection and theres already so many layers. I feel that outweighs the organization benefit that comes from grouping timeseries tests together

ricardoV94 replied:
I don't follow. You just need to import the test class from this file, no?

canyon289 commented on pymc/tests/test_distributions.py:
```diff
@@ -2608,6 +2608,15 @@ def test_interpolated_transform(self, transform):
+            pm.GaussianRandomWalk,
+            R,
+            {"mu": R, "sigma": Rplus, "steps": Nat},
+            lambda value, mu, sigma: sp.norm.logpdf(value, mu, sigma).cumsum(),
```
How can I use this test with **values**, not just a single value?

**

canyon289 replied:
Note this is currently failing

ricardoV94 replied:
This test will try a bunch of values from `R`

ricardoV94 replied:
But I think it should be a `Vector` of `R`, not just a scalar. I think there are some created in `test_distributions` that you can have a look at

ricardoV94 replied:
I would put this test in the other module with the rest of the timeseries tests

OriolAbril commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +36,137 @@
+
+        Parameters
+        ----------
+        rng: np.random.RandomState
```
does this only take randomstate and not generators? see https://numpy.org/doc/stable/reference/random/

OriolAbril commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +36,137 @@
+        ----------
+        rng: np.random.RandomState
+           Numpy random number generator
+        mu: np.ndarray
```
this should be `array_like` instead. It won't match the type hints but it will match the actual accepted types, they are passed as is to the numpy function

OriolAbril commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +36,137 @@
+
+        Returns
+        -------
+        np.ndarray
```
and this should be `ndarray` only. `np` means nothing to sphinx and we have an alias set for ndarray if used on its own. See https://pymc-data-umbrella.xyz/en/latest/sprint/docstring_tutorial.html#parameters

OriolAbril commented on pymc/distributions/timeseries.py:
```diff
@@ -176,127 +310,6 @@ def logp(self, value):
-
-    Parameters
-    ----------
-    mu : tensor_like of float, default 0
```
these (also the ones below) type descriptions are much better than the ones currently proposed, they use `tensor_like` which is rendered as a link to the right definition on the glossary and the default/optional parameter info.

ricardoV94 commented:
I didn't mean to remove init altogether! Just to not account for the initial value in the logp

canyon289 commented:
> I didn't mean to remove init altogether! Just to not account for the initial value in the logp
Ah no worries, Ill drop the commits theyre removed then. I get what youre saying now


@OriolAbril thanks for the comments I'll address them all

canyon289 commented:
Moved the infer steps commit to this PR https://github.com/pymc-devs/pymc/pull/5541

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -58,28 +61,25 @@ def test_grw_rv_op_shape(self, kwargs, expected):
+        # `at.diff` is currently broken with constants
+        test_vals = [0, 1, 2]
+        vals = at.vector("vals")
         mu = 1
```
TODO For myself. I'll remove this since its not used anymore

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -88,16 +88,15 @@ def test_grw_inference(self):
-            with pytest.raises(TypeError) as err:
-                trace = pm.sample()
+            # Workaround for bug in `at.diff` when data is constant
+            obs_data = pm.MutableData("obs_data", obs)
```
@ricardoV94 Just curious how this gets around the bug. What does the MutableData object change in the Aesara logic?

ricardoV94 replied:
`MutableData` uses a `SharedVariable` whose size is allowed to change. The Aesara bug happens when you have a constant whose size is not allowed to change anymore and it wrongly guesses what the new output size will be. It "forgets" that it is (size-1). I'll see if I can patch it today in Aesara

canyon289 replied:
If you change can you cc on the pr, or link here. it'll add to the learning

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -142,14 +145,37 @@ class GaussianRandomWalk(distribution.Continuous):
 
     rv_op = gaussianrandomwalk
 
+    def __new__(cls, name, mu=0.0, sigma=1.0, init=None, steps: int = 1, **kwargs):
```
Why is overriding `__new__` method needed?

ricardoV94 replied:
`__new__` is the one called when you create a "named" distribution like `pm.Normal("x"...)` instead of calling `.dist()` directly as in `pm.Normal.dist(...)`. This is where it makes sense to check that `init` is not a variable already registered in the model (to avoid double counting)

canyon289 replied:
what is happening in `__new__` makes sense, to be more precise why do  we use `__new__` instead of `__init__`?

ricardoV94 replied:
Because these are not real classes, we never initialize pymc distributions only the underlying `rv_op` RandomVariables.

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -174,11 +200,13 @@ def logp(
         """
 
         # Calculate initialization logp
+        init_logp = logprob.logp(init, value[..., 0])
```
what happened to double counting init? Is that not an issue anymore?

ricardoV94 replied:
No, we added a check in `__new__` that `init` is not part of the model variables so it's logp is not counted anywhere else, only here. See my comment to your question above

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -107,10 +109,10 @@ def rng_fn(
             steps_size = (*size, steps)
 
-        init_val = rng.normal(init, sigma, size=init_size)
+        init = np.reshape(init, init_size)
```
Why is no random draw needed here? Is it because we defer to the random method of the init dist?

ricardoV94 replied:
Yes, we assume init already has the right number of random draws (that's why we resize it in `dist`)

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -12,20 +12,118 @@
+        "steps,size,expected",
+        (
+            (1, None, (2,)),
+            (2, 1, (1, 3)),
```
This fails now. Obviously because we added shapes, but does this mean now that  shape from params function needs to be updated to account for size

https://github.com/pymc-devs/pymc/pull/5298/files#diff-5fece42aa9647732b360b6c0e13de089a8408c68ae38b650fba38a1c40322229R52

<img width="748" alt="image" src="https://user-images.githubusercontent.com/7213793/156865913-a27eecd8-3702-4865-82f8-8b66cdd5bf7f.png">

ricardoV94 replied:
Have you updated Aesara? The meaning of size for multivariate dists changed and should now mean this test is correct.

canyon289 replied:
That  was the issue thanks

canyon289 commented:
New question  `check_pymc_params_match_rv_op` because expected_value is None, which is a test input, but the actual_value is a shape 15 random draw which I dont quite understand


<img width="1357" alt="image" src="https://user-images.githubusercontent.com/7213793/156890088-54152f2b-a81b-42eb-af05-fc29844de727.png">

ricardoV94 commented:
> New question  `check_pymc_params_match_rv_op` because expected_value is None, which is a test input, but the actual_value is a shape 15 random draw which I dont quite understand
> 
> 
> <img width="1357" alt="image" src="https://user-images.githubusercontent.com/7213793/156890088-54152f2b-a81b-42eb-af05-fc29844de727.png">
> 

Is that `init`? Remember that `init` is a random variable with the same size as the GRW. You might want to use `init=pm.Constant(0)` so that all the expected values are `0`

canyon289 commented:
I tried that but the issue is that because were not in a model context it fails :/

i'm also confused because the GRW isn't shape 15 with those inputs, it should be shape 4, since there only 4 steps. Need to dig in to figure out where 15 comes from

<img width="1545" alt="image" src="https://user-images.githubusercontent.com/7213793/156890731-7c8e913c-e96f-4f99-9a90-372d45f7e8b3.png">

ricardoV94 commented:
> I tried that but the issue is that because were not in a model context it fails :/
> 
> i'm also confused because the GRW isn't shape 15 with those inputs, it should be shape 4, since there only 4 steps. Need to dig in to figure out where 15 comes from
> 
> <img width="1545" alt="image" src="https://user-images.githubusercontent.com/7213793/156890731-7c8e913c-e96f-4f99-9a90-372d45f7e8b3.png">
> 

Sorry, I meant `Constant.dist(0)`, but yeah seems like a different problem

canyon289 commented:
Something is going on with size (or shape?) thats  confusing me. Ill keep digging in to figure it out.

This is just an overall note, in section of the codebase when one thing is changed it has a lot of side effects that are hard to determine if not aware of the whole system. Trying to learn it I promise, appreciate the continued help


<img width="1534" alt="image" src="https://user-images.githubusercontent.com/7213793/156891007-84421ed5-0630-40c3-9b14-a71f7a73c011.png">

ricardoV94 commented:
Ahh... default size might be 15. Try to add an attribute `size=None` in the class

canyon289 commented:
edit; Oh in the class, ok

<img width="1290" alt="image" src="https://user-images.githubusercontent.com/7213793/156891194-cfb48a63-fb12-43e5-bff0-484b853e44d6.png">

canyon289 commented:
> Ahh... default size might be 15. Try to add an attribute `size=None` in the class

<img width="1442" alt="image" src="https://user-images.githubusercontent.com/7213793/156891392-dd4f9bd4-b5d3-4ebd-a6c5-44aa6fe8ba64.png">

canyon289 commented:
Here's the new place I'm stick.

We can't use assert almost equal here since the value passed is not a numeric type. Should i add an if then statement to check for equality? I dislike conditionals in tests, though this is more of a framework so I suppose its ok


<img width="1114" alt="image" src="https://user-images.githubusercontent.com/7213793/156938270-9242f2eb-3077-4ee1-9e02-12cfae16a210.png">

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -12,20 +12,156 @@
 
-# pytestmark = pytest.mark.usefixtures("seeded_test")
-pytestmark = pytest.mark.xfail(reason="Timeseries not refactored")
+from .test_distributions_random import BaseTestDistributionRandom
```
@ricardoV94 I moved the test class but i really dont think we should do this. i somewhat strongly feels test should be self contained in one module as it makes it easier to read and see whats going on

ricardoV94 replied:
I think the tests files should mirror the original codebase. So they should be in a file with the same name as where the functions/classes that it tests are defined. Ideally, timeseries tests should be in test/distributions/test_timeseries.py. This is how Aesara does it and its much easier to find what tests relate to what code, and also to spot gaps in our tests.

On the other hand, if you follow your logic you shouldn't have any test_timeseries.py at all. The random tests would go into test_distributions_random.py, the logp into test_distributions.py, and so on. Why would you have this test file at all?

ricardoV94 replied:
Unrelated to the discussion, you should use absolute imports not relative one (that is the `.`)

canyon289 replied:
> On the other hand, if you follow your logic you shouldn't have any test_timeseries.py at all. The random tests would go into test_distributions_random.py, the logp into test_distributions.py, and so on. Why would you have this test file at all?

Yep agree, its already a bit confusing because we dont split per the convention of tests mirroring the codebase, we have two conventions going on at once. So if we go that way i think we should merge `test_distributions_random` and `test_distributions_logprob` etc but anyhow that is out of the scope of this pr :)

I'll update the relative import

ricardoV94 replied:
Opened up discussion in https://github.com/pymc-devs/pymc/discussions/5579

canyon289 replied:
Thanks

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,185 @@
+        init_logp = logprob.logp(init, value[..., 0])
+
+        # Make time series stationary around the mean value
+        stationary_series = at.diff(value, axis=-1)
```
I think we can do it manually instead of using `at.diff` to avoid the bug. Not only is it faster, because `diff` does not have a `C` implementation, but also the gradient will also work for `value` that have more than 1 dimension, whereas the gradient of `at.diff` would raise a `NotImplementedError`:

```suggestion
        stationary_series = value[..., 1:] - value[..., :-1]
```

ricardoV94 commented on pymc/tests/test_distributions_random.py:
```diff
@@ -418,29 +315,6 @@ def seeded_numpy_distribution_builder(dist_name: str) -> Callable:
     )
 
 
-class TestFlat(BaseTestDistributionRandom):
```
Was the `TestFlat` removed by accident?

canyon289 replied:
Yes readded

canyon289 commented:
> Here's the new place I'm stick.
> 
> We can't use assert almost equal here since the value passed is not a numeric type. Should i add an if then statement to check for equality? I dislike conditionals in tests, though this is more of a framework so I suppose its ok
> 
> <img alt="image" width="1114" src="https://user-images.githubusercontent.com/7213793/156938270-9242f2eb-3077-4ee1-9e02-12cfae16a210.png">

@ricardoV94 I think this is the only thing holding up this PR now. Let me know if I should extend the framework

canyon289 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -12,20 +12,155 @@
+    size = None
+
+    pymc_dist = pm.GaussianRandomWalk
+    pymc_dist_params = {"mu": 1.0, "sigma": 2, "init": pm.Constant.dist(0), "steps": 4}
```
@ricardoV94 regarding the earlier comment regarding test failures its because we did this. Init doesn't support integer arguments anymore given it must be scalar distribution.

ricardoV94 replied:
What happens if you omit init?

canyon289 replied:
See https://github.com/pymc-devs/pymc/pull/5298#issuecomment-1065356038

canyon289 commented:
> What happens if you omit `init`?

@ricardoV94 Then we get an arg length mismatch exception
<img width="1092" alt="image" src="https://user-images.githubusercontent.com/7213793/157924708-d755163b-f2c0-4ed2-8ad9-9cf72ac0709c.png">

Here's what happens if its a scalar as all other tests are.

<img width="1075" alt="image" src="https://user-images.githubusercontent.com/7213793/157924897-7b52f391-c782-4b96-9a46-b4892b911980.png">

ricardoV94 commented:
I see, I would not extend the BaseTest, but override the method in this class

canyon289 commented:
> I see, I would not extend the BaseTest, but override the method in this class

Rather than override I may introduce a new check so its less confusing, but either way it seems were in agreement a new set of logic is needed to handle this case. Thanks!

canyon289 commented:
@ricardoV94 any idea why this is failing now? https://github.com/pymc-devs/pymc/runs/5531778572?check_suite_focus=true#step:7:1765

I get that ndim support is out of range, but since nothing has changed in the distribution itself and `ndim_supp` should be once it seems strange. Is `rep_param_idx` wrong now?

```        if param_shapes is not None:
            ref_param = param_shapes[rep_param_idx]
>           return (ref_param[-ndim_supp],)
E           IndexError: tuple index out of range
```

Edit: Nevermind figured it out, its because the updated name is `_supp_shape_from_params`

canyon289 commented:
All tests pass except precommit :tada:.

@ricardoV94 aside from precommit any other changes youd like me to make?

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,185 @@
+    def _supp_shape_from_params(self, dist_params, reop_param_idx=0, param_shapes=None):
+        steps = dist_params[3]
+
+        # TODO: Ask ricardo why this is correct. Isn't shape different if size is passed?
```
```suggestion
```
Hopefully the new method name makes it obvious that it should return the support shape ignoring size

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,185 @@
+            steps_size = steps
+
+        # TODO: Ask Ricardo how size is known to be a tuple? Its int above
+        # If size is None then the returned series should be (size, 1+steps)
```
Is this TODO still relevant?

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -2616,6 +2616,16 @@ def test_interpolated_transform(self, transform):
+    def test_grw(self):
+        self.check_logp(
+            pm.GaussianRandomWalk,
+            R,
```
Does this work?
```suggestion
            Vector(R, 10)
```

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -2616,6 +2616,16 @@ def test_interpolated_transform(self, transform):
+            pm.GaussianRandomWalk,
+            R,
+            {"mu": R, "sigma": Rplus, "steps": Nat},
+            lambda value, mu, sigma: sp.norm.logpdf(value, mu, sigma).cumsum().sum(),
```
This has to be updated to include `init`

canyon289 replied:
I thought init was not included in the logp calc though and is accounted for elsewhere?  If thats the case why would we need to include init to verify the logp calc?

ricardoV94 replied:
It is included again since the last couple of commits.

canyon289 replied:
How can the logp for init be included here? I'm confused because the init distribution is declared elsewhere and passed to GRW, but the logp for the init dist is calculated within the grws logup. 

That all makes sense up until this fixture where its not clear how to properly do that in the text context. Any ideas?

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -3407,7 +3417,8 @@ def test_no_warning_logp(self):
             pm.Exponential.dist(1),
-            pm.MvNormal.dist(np.ones(3), np.eye(3)),
+            # For some reason this test is running when TimeSeries test is run
+            # pytest.mark.xfail(pm.MvNormal.dist(np.ones(3), np.eye(3)),)
```
Weird. Is this locally, or also on the Github CI?

canyon289 replied:
Not sure, I'll uncomment when PR is ready since it burns up a lot of CI time trying to figure out

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,185 @@
+                # If not explicit, size is determined by the shape of mu and sigma
+                mu_ = at.broadcast_arrays(mu, sigma)[0]
+                init = change_rv_size(init, mu_.shape)
+
```
We need to tell Aeppl that it's fine to ignore this random variable, because the logp is taken care of somewhere else.
Otherwise there will be a warning

```suggestion

        init.tag.ignore_logprob = True

```

canyon289 replied:
If I understand this change correctly its removing `init = change_rv_size(init, mu_.shape` and adding `init.tag.ignore_logprob = True`? I have questions but will make sure I know whats happening first :)

ricardoV94 replied:
No, the change is supposed to add that line after the if/else block. It should not remove anything

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,187 @@
+                mu_ = at.broadcast_arrays(mu, sigma)[0]
+                init = change_rv_size(init, mu_.shape)
+
+        # Ignores logprob of init var because that's captured somewhere else
```
```suggestion
        # Ignores logprob of init var because that's accounted for in the logp method
```

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -3410,7 +3420,9 @@ def test_no_warning_logp(self):
         [
             pm.Exponential.dist(1),
-            pm.MvNormal.dist(np.ones(3), np.eye(3)),
+            pytest.mark.xfail(
```
this shouldn't have changed in your PR AFAICT

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -2619,6 +2619,16 @@ def test_interpolated_transform(self, transform):
             assert not np.isfinite(m.compile_logp()({"x": -1.0}))
             assert not np.isfinite(m.compile_logp()({"x": 11.0}))
 
+    @pytest.mark.skip("Unsure how to use this test fixture")
```
Should also go to `test_timeseries.py`

canyon289 commented:
FYI Although I know its good form to update the type hints for timeseries I'm not inclined to do it in this pr. What are the instructions for updating the type hint ignore and deferring that to another PR?

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,187 @@
+    @classmethod
+    def dist(
+        cls, mu=0.0, sigma=1.0, init=None, steps: int = 1, size=None, **kwargs
+    ) -> RandomVariable:
```
```suggestion
    ) -> TensorVariable:
```

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +35,187 @@
+                and isinstance(init.owner.op, RandomVariable)
+                and init.owner.op.ndim_supp == 0
+            ):
+                raise TypeError("init must be a scalar distribution variable")
```
```suggestion
                raise TypeError("init must be a univariate distribution variable")
```

ricardoV94 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -11,21 +11,94 @@
+            pm.GaussianRandomWalk,
+            td.Vector(td.R, 10),
+            {"mu": td.R, "sigma": td.Rplus, "steps": td.Nat},
+            lambda value, mu, sigma: stats.norm.logpdf(value, mu, sigma).cumsum().sum(),
```
This expression is wrong:
```suggestion
            lambda value, mu, sigma, init, steps: stats.norm.logpdf(np.concatenate([[value[0]], np.diff(value)], axis=-1)), mu, sigma).sum(),
```
Probably better to use a function than a lambda, to break that large one-liner

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -32,6 +36,191 @@
+        """
+
+        if steps is None or steps < 1:
+            raise ValueError("Steps must be None or greater than 0")
```
I think we can allow steps to be zero, in which case the distribution is just init

canyon289 replied:
Im inclined not to do this because i dont think anyone would want to do this on purpose and if they do it accidentally this will help them catch the mistake

ricardoV94 replied:
I disagree. Even for educational purposes it might be useful to show such a progression from init to first steps. 

We can make steps a required argument (I don't remember why we reverted on that decision) which will force you to choose steps. That's probably safer than our current default of 1, specially since its a new argument.

canyon289 replied:
Ok i'll open an issue to track for another PR so we can get this one merged

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -2619,6 +2619,21 @@ def test_interpolated_transform(self, transform):
             assert not np.isfinite(m.compile_logp()({"x": -1.0}))
             assert not np.isfinite(m.compile_logp()({"x": 11.0}))
 
+    def test_grw(self):
```
Moved this back here, because of the inheritance issues that you found @canyon289. We will eventually refactor this test class and split the tests across different modules, but there's no reason to do it in this PR

canyon289 replied:
Great marking for another issue

ricardoV94 commented on pymc/tests/test_distributions_timeseries.py:
```diff
@@ -11,21 +11,85 @@
+
+    # TODO: Move this out of this test class, not related, makes more sense together with 
+    # the other logp tests
+    def test_grw_inference(self):
```
This is a pretty heavy test. Can we do with less steps, and a single chain? We can make the check more tolerant

canyon289 replied:
I reduced the steps, I think its only one chain as is?

canyon289 replied:
Oh do you mean one MCMC chain?

ricardoV94 replied:
Yeah one MCMC chain

ricardoV94 commented on pymc/tests/test_distributions.py:
```diff
@@ -2619,6 +2619,22 @@ def test_interpolated_transform(self, transform):
             assert not np.isfinite(m.compile_logp()({"x": -1.0}))
             assert not np.isfinite(m.compile_logp()({"x": 11.0}))
 
+    def test_grw(self):
```
```suggestion
    def test_gaussianrandomwalk(self):
```

canyon289 commented:
All tests pass which is great. I think precommit is not failing locally for me, but nonetheless Ill try and figure out what the issue is and also learn how to read precommit

canyon289 committed:
Add v4 grw and tests

canyon289 committed:
Remove time series from mypy checks for now

canyon289 commented:
Alright i propoose we merge even with precommit failinng because its no worse than the current state of precommit. 

In other prs we can get all the other stuff done. I already cleaned up the git history

https://github.com/pymc-devs/pymc/pull/5673

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,189 @@
+        if "shape" in kwargs.keys():
+            shape = kwargs["shape"]
+        else:
+            shape = None
```
```suggestion
        shape = kwargs.get("shape", None)
```

ricardoV94 committed:
Make steps required explicitly and check it is a scalar

ricardoV94 committed:
Move inference test out of unrelated class

ricardoV94 committed:
Fix init dist and add tests

ricardoV94 committed:
Fix docstrings

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,192 @@
+    dtype = "floatX"
+    _print_name = ("GaussianRandomWalk", "\\operatorname{GaussianRandomWalk}")
+
+    def make_node(self, rng, size, dtype, mu, sigma, init, steps):
```
Curious why is this function needed now? Is this check already not covered here?

https://github.com/pymc-devs/pymc/pull/5298/files#diff-5fece42aa9647732b360b6c0e13de089a8408c68ae38b650fba38a1c40322229R102

ricardoV94 replied:
That line is a value check, this is a shape check (to confirm it is a scalar)

canyon289 replied:
Why this function then? I'm assuming make_node is called when the aesara graph is being made?

In other words why cant this logic be moved to where the value check is? Not saying it should be, just trying to understand :)

ricardoV94 replied:
This check happens immediately when the variable is created, while that other one only when/if the variable is evaluated.

We don't want to allow the user to even initialize such variables, whereas we can't do that for the value of the steps, because it might be symbolic and not known until the function is evaluated.

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -143,17 +148,19 @@ class GaussianRandomWalk(distribution.Continuous):
     rv_op = gaussianrandomwalk
 
-    def __new__(cls, name, mu=0.0, sigma=1.0, init=None, steps: int = 1, **kwargs):
+    def __new__(cls, name, mu=0.0, sigma=1.0, init=None, steps=None, **kwargs):
```
I think I asked this question before but do `__new__` and `dist` have to have almost the same signature (except for size)?

ricardoV94 replied:
They can have the same signature. I don't know if other distributions usually show or hide size in kwargs. Only thing we care about here is getting `init`

canyon289 replied:
Is it ever the case theyll have vastly different signatures?

ricardoV94 replied:
`__new__` is only used internally. Usually we just need to take some arguments for checks like the one here. Otherwise we leave it unchanged/undefined.

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -104,19 +104,20 @@ def rng_fn(
         if size is None:
-            init_size = 1
-            steps_size = steps
+            bcast_shape = np.broadcast_shapes(
```
What does broadcast shape here do specifically that the code before was not doing?

ricardoV94 replied:
It makes sure that when `init` is larger than mu or sigma, the distribution has the right shape, i.e., `(*init.shape, nsteps)`

canyon289 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,192 @@
+
+        # If size is None then the returned series should be (1+steps,)
+        if size is None:
+            bcast_shape = np.broadcast_shapes(
```
```suggestion
            # When init is larger than mu or sigma ensure the distribution has the right shape, (*init.shape, nsteps)
             bcast_shape = np.broadcast_shapes(
```

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,192 @@
+
+    def make_node(self, rng, size, dtype, mu, sigma, init, steps):
+        steps = at.as_tensor_variable(steps)
+        if not steps.ndim == 0 or not steps.dtype.startswith("float"):
```
```suggestion
        if not steps.ndim == 0 or not steps.dtype.startswith("int"):
```

whoops

canyon289 committed:
Update pymc/distributions/timeseries.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

canyon289 committed:
Update pymc/distributions/timeseries.py

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,193 @@
+        if steps < 1:
+            raise ValueError("Steps must be greater than 0")
+
+        # If size is None then the returned series should be (1+steps,)
```
```suggestion
        # If size is None then the returned series should be (*implied_dims, 1+steps)
```

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,193 @@
+
+        # If size is None then the returned series should be (1+steps,)
+        if size is None:
+            # When init is larger than mu or sigma ensure the distribution has the right shape, (*init.shape, nsteps)
```
```suggestion
            # broadcast parameters with each other to find implied dims
```

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,193 @@
+            )
+            dist_shape = (*bcast_shape, int(steps))
+
+        # If size is None then the returned series should be (size, 1+steps)
```
```suggestion
        # If size is None then the returned series should be (*size, 1+steps)
```

ricardoV94 reviewed (commented):
small tweak to comments

ricardoV94 commented:
There's some indentation error going on

canyon289 committed:
Update timeseries

canyon289 committed:
Update pymc/distributions/timeseries.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

canyon289 committed:
Update pymc/distributions/timeseries.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

canyon289 committed:
Update pymc/distributions/timeseries.py

Co-authored-by: Ricardo Vieira <28983449+ricardoV94@users.noreply.github.com>

canyon289 commented:
Looks like the init dist is causing some test failure. I can look into this weekend, next 72 hours or so

https://github.com/pymc-devs/pymc/runs/5788430816?check_suite_focus=true#step:7:411
![image](https://user-images.githubusercontent.com/7213793/161277146-2c579d4f-af3c-4ada-8064-0bb00ecf2f6a.png)

ricardoV94 commented on pymc/distributions/timeseries.py:
```diff
@@ -33,6 +38,193 @@
+    dtype = "floatX"
+    _print_name = ("GaussianRandomWalk", "\\operatorname{GaussianRandomWalk}")
+
+    def make_node(self, rng, size, dtype, mu, sigma, init, steps):
```
This method should return super().make_node(...). Not sure what happened with my changes xD. Must have changed and not retested locally

Here is a reference distribution doing the correct thing: https://github.com/pymc-devs/pymc/blob/741d45540016ab52a15ad5bff521614b6ae1d835/pymc/distributions/multivariate.py#L291

canyon289 replied:
Ok great, ill take a look and as always Ill have questions, in the spirit of learning :)

canyon289 replied:
Thanks buddy! That did the trick, and I really appreciate you showing me what to do

canyon289 committed:
Add call to super for make node

canyon289 commented:
Alright well this passes except black which I think is broken, and mypy which Im pretty sure is broken, and flaky codecov.  Should I just merge and those get fixed up in other prs?
